skrifa/src/patchmap.rs

```diff
@@ -3,223 +3,295 @@
-                // TODO(garretrieger): need to error out instead of ignoring.
+        if let Some(largest_tag) = largest_tag {
+            if *tag <= largest_tag {
+                // Out of order or duplicate tag, skip this record.
+                next_tag = tag_it.next();
                 continue;
             }
+        }
 
-            let Some(new_entry) = entries.get_mut(m.new_entry_index as usize) else {
-                // TODO(garretrieger): need to error out instead of ignoring.
-                continue;
-            };
+        largest_tag = Some(*tag);
 
-            new_entry.codepoints.union(&mapped_codepoints);
-            new_entry.feature_tags.insert(m.feature_tag);
-            new_entries.insert(m.new_entry_index);
+        let entry_count = record.entry_map_count().get();
+        if *tag < record.feature_tag() {
+            next_tag = tag_it.next();
+            continue;
         }
 
-        present_entries.union(&new_entries);
-    }
+        for i in 0..entry_count {
+            let index = i + cumulative_entry_map_count;
+            let byte_index = (index * field_width * 2) as usize;
+            let data = FontData::new(&feature_map.entry_map_data()[byte_index..]);
+            let mapped_entry_index = record.first_new_entry_index().get() + i;
+            let record = EntryMapRecord::read(data, max_entry_index)?;
+            let first = record.first_entry_index().get();
+            let last = record.last_entry_index().get();
+            if first > last
+                || first > max_glyph_map_entry_index
+                || last > max_glyph_map_entry_index
+                || mapped_entry_index <= max_glyph_map_entry_index
+                || mapped_entry_index > max_entry_index
+            {
+                // Invalid, continue on
+                continue;
+            }
 
-    /// Produce a mapping from each glyph id to the codepoint(s) that map to that glyph.
-    fn glyph_to_unicode_map<'a>(font: &impl TableProvider<'a>) -> HashMap<GlyphId, IntSet<u32>> {
-        let charmap = Charmap::new(font);
-        let mut glyph_to_unicode = HashMap::<GlyphId, IntSet<u32>>::new();
-        for (cp, glyph) in charmap.mappings() {
-            glyph_to_unicode
-                .entry(glyph)
-                .or_insert_with(IntSet::<u32>::empty)
-                .insert(cp);
+            if entries.intersects_range(first..=last) {
+                entries.insert(mapped_entry_index);
+            }
         }
-        glyph_to_unicode
+        next_tag = tag_it.next();
     }
 
-    fn apply_uri_template(uri_template: &str, _entry_index: usize) -> String {
-        // TODO(garretrieger): properly implement this, may deserve to go into it's own module.
-        uri_template.to_string()
-    }
+    Ok(())
+}
 
-    // TODO(garretrieger): add template application method that takes a string entry id (will be needed for format 2)
+#[derive(Clone, Eq, PartialEq, Debug, Hash, Copy)]
+pub enum PatchEncoding {
```

Comment:
A doc comment with a brief description and a link to the spec would be nice here.